lib/optimizer/src/Transformer/ServerSideRendering.php

```diff
@@ -620,4 +708,274 @@ private function isAmpExperimentUsed(DOMElement $element)
+            // Therefore, it should simply be stripped, without producing media queries.
+            // @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img#attr-sizes
+            return '';
+        }
+
+        return $this->extractAttributeCss(
+            $document,
+            $element,
+            $attribute,
+            '#__ID__{width:%s}',
+            '@media %s{#__ID__{width:%s}}'
+        );
+    }
+
+    /**
+     * Extract the custom CSS styling from a 'heights' attribute.
+     *
+     * __heights__ (AMP-specific)
+     * "The value of this attribute is a sizes expression based on media expressions, similar to the sizes attribute on
+     * img tags but with two key differences:
+     * - The value applies to the height, not the width of the element.
+     * - Percent values are allowed. A percent value indicates the percent of the element's width. For example, a value
+     *   of 80% indicates that the height of the element will be 80% of the element's width."
+     *
+     * @see https://amp.dev/documentation/guides-and-tutorials/learn/common_attributes/#heights
+     *
+     * @param Document   $document  Document containing the element to adapt.
+     * @param DOMElement $element   Element to adapt.
+     * @param DOMAttr    $attribute Attribute to be extracted.
+     * @return string Extract custom CSS styling.
+     */
+    private function extractHeightsAttributeCss(Document $document, DOMElement $element, DOMAttr $attribute)
+    {
+        $this->customSizerStyles[$document->getElementId($element)] = '';
+
+        return $this->extractAttributeCss(
+            $document,
+            $element,
+            $attribute,
+            '#__ID__>:first-child{padding-top:%s}',
+            '@media %s{#__ID__>:first-child{padding-top:%s}}'
+        );
+    }
+
+    /**
+     * Extract the custom CSS styling from an attribute and turn into a templated CSS style string.
+     *
+     * @param Document   $document        Document containing the element to adapt.
+     * @param DOMElement $element         Element to adapt.
+     * @param DOMAttr    $attribute       Attribute to be extracted.
+     * @param string     $mainStyle       CSS template for the main style.
+     * @param string     $mediaQueryStyle CSS template for a media query style.
+     * @return string Extract custom CSS styling.
+     */
+    private function extractAttributeCss(Document $document, DOMElement $element, DOMAttr $attribute, $mainStyle, $mediaQueryStyle)
+    {
+        if (empty($attribute->nodeValue)) {
+            return '';
+        }
+
+        $sourceSizes = explode(',', $attribute->nodeValue);
+        $lastItem    = trim(array_pop($sourceSizes), self::CSS_TRIM_CHARACTERS);
+
+        if (empty($lastItem)) {
+            throw InvalidHtmlAttribute::fromAttribute($attribute->nodeName, $element);
+        }
+
+        $customCss = sprintf($mainStyle, $lastItem);
+
+        foreach ($sourceSizes as $sourceSize) {
```

Comment:
Do we need to iterate this in reverse order? The spec says:

> The user agent will pick a width from the sizes attribute, using the first item with a <media-condition> (the part in parentheses) that evaluates to true

This means, we have to reverse the order the media queries in CSS to emulate this behavior (the last definition has precedence).

```
<amp-img id="img1" sizes="(min-width: 200px) 200px, (min-width: 320px) 320px, 100vw"></amp-img>
```

Should result in:

```
#elem {
  width: 100vw;
}
@media  (min-width: 320px) {
  #elem {
    width: 320px;
  }
}
@media (min-width: 200px) {
  #elem {
    width: 200px;
  }
}
```

Or am I missing something?

Reply:
Yes, you're right, the CSS evaluates in the opposite order.

However, considering this, note that the `sizes` example above is actually broken, it would never pick the middle option.

Reply:
This was missed in a previous follow-up PR. New follow-up PR to fix this: https://github.com/ampproject/amp-wp/pull/5263